README

                             prima.py 0.3
			     ************

USAGE
=====

This is a wrapper script for Maurits Haverkort's
‘SpaghettiPrimavera.f90’.  It produces the same kind of plots but
should provide a simple interface.  Like ‘SpaghettiPrimavera.f90’, it
needs a ‘case.qtl’ and a ‘case.klist_band’ file; furthermore it reads
a ‘case.struct’ if present.

If it exists, ‘case.qtl[up|dn]_band’ is preferred over
‘case.qtl[up|dn]’.  (Wien2k does not use ‘.qtl_band’ but it may be
convenient to rename a ‘.qtl’ file intended for band structures.)

See ‘prima.py -h’ and ‘prima.py’ (executed in a suitable ‘case’
directory) for usage hints.


An Example
----------

Suppose we are interested in the t2g and eg contributions of the first
inequivalent atom in the struct.  A typical command line might be

   prima.py 1:D-t2g:5,0,0:3 1:D-eg:0,0,5:3 -o prima.ps

if that atom has ‘ISPLIT=2’.  If instead it has ‘ISPLIT=8’, the
command would read:

   prima.py 1:DXY&DXZ&DYZ:5,0,0:3 1:DX2Y2&DZ2:0,0,5:3 -o prima.ps

which is an abbreviated form of

   prima.py 1:DXY:5,0,0:3 1:DXZ:5,0,0:3 1:DYZ:5,0,0:3 \
            1:DX2Y2:0,0,5:3 1:DZ2:0,0,5:3 -o prima.ps

Each of the non-option arguments names an atom, an orbital, a color, a
thickness increment, and a description for the (optional) legend;
these tokens are separated by colons (‘:’), and each of them may be
empty.


Specification of Atoms and Orbitals
-----------------------------------

The ampersand (‘&’) combines two orbitals or two sites.  Since it must
be escaped in most shells, two plus signs (‘++’) may be used instead
(a single ‘+’ appears in some orbital descriptions in ‘.qtl’ files).

For both the atom and the orbital specification, you can use a number
(in the sequence of ‘case.struct’ or ‘case.qtl’) or a name (labels as
given in the same files).  Furthermore, the following special atom
names are accepted:

  all  => all atoms
  istl => the interstitial
  Z$N  => all atoms with atomic number $N
  $Sy  => the atomic symbol expands to all atoms of that element

The atomic symbol applies only if you have several inequivalent atoms
of the same element.  As long as you follow the Wien2k convention for
the struct labels (atomic symbol in the first two places, a number in
the third), there can be no ambiguity with these special cases.

All atom and orbital specifications are case insensitive.


Colors and Normalization
------------------------

Colors can be named in three ways: as a comma-separated triple “r,g,b”
of red, green, blue values (nominally ∈ [0, 1]), in hexadecimal
notation à la HTML “#RRGGBB”, or one of the 147 color names defined in
SVG <http://www.w3.org/TR/SVG11/types.html#ColorKeywords>.

The QTL values from Wien2k, when the interstitial is included, are
normalized to 1, but how much of this weight is contained in a
particular muffin-tin sphere depends on a number of things.  This
leads to a “normalization problem”; since there is no general
solution, prima.py simply multiplies the colors and thickness
increments by the QTL values and adds them.  Therefore, it can be
useful to have r,g,b > 1.  To this end, the color may be prefixed by a
factor: “$f × $color”.

In general, colored spaghetti are not a quantitative tool, and care
must be taken in their interpretation, especially concerning the
normalization and mixing of colors.


Configuration Files
-------------------

The option -C|--config-file may be used to read options from a file in
addition to the command line.  Each line in the config file(s) should
contain one option and, if applicable, its argument, separated by
whitespace.  Any option valid on the command line may be used, leading
dashes may be omitted.  For instance, if, by default, you want to
suppress the minor ticks, add a legend, and output to
‘$case_prima.ps’, then put the following into ‘~/.prima’

   minor-ticks	0
   write-legend
   out-suffix   _prima.ps

and add

   alias prima.py="prima.py -C ~/.prima"

to your ‘~/.bashrc’ (mutatis mutandis for other shells).  Beside a
global config file, it can also be useful to have per-directory
configuration.  This could be implemented by

  alias prima.py='prima.py $(test -e .prima && echo "-C .prima")'

Parsing of the config files is primitive.  Lines are skipped if they
are empty or start with a hash (‘#’).  Leading whitespace and
whitespace between option and argument is stripped, but trailing
whitespace is preserved.


INSTALLATION
============

prima.py needs Python 2, NumPy, F2PY (which is now part of NumPy), and
a modern Fortran compiler.  I tested it with the following versions:

Python 	 2.7.3
NumPy  	 1.6.1
F2PY   	 2
gfortran 4.6.3

Running ‘make’ in the ‘prima’ directory should produce the file
‘sppv.so’.  Python must be able to find this file, and the file
‘webcolors.py’.  This is easiest if you create a symlink to prima.py
in some directory in your $PATH, or add the ‘prima’ directory to your
$PATH.


COMMENTS and SUGGESTIONS
========================

about prima.py should go to <elias.assmann@gmail.com>.

The author of the original SpaghettiPrimavera.f90 is
Maurits W. Haverkort <Maurits@physics.ubc.ca>.


ACKNOWLEDGEMENTS
================

Thanks to Maurits Haverkort for letting me build upon his code.

Hex colors and color names are handled by the ‘webcolors’ module,
written by James Bennett.

Thanks to Peter Blaha and Manuel Zingl for valuable comments on
previous versions.


NEWS
Version 0.3 (2015-09-xx):

 * add option --updn to combine spins in one plot

 * read Fermi energy from ‘scf’ rather than ‘scf2’; do not require it
   unless needed

 * change delimiter for combined atoms/orbitals to '++|&' to avoid
   clash with orbital names that contain ‘+’

 * new options --out-suffix, --config-file, --verbose

 * make atoms and orbitals case insensitive

 * colors may be also given in HTML-style hex notation or color names;
   and may be prefixed by a factor, “F × COLOR” for scaling


Version 0.2 (2014-09-01):

 * add support for ‘x qtl’-generated files

 * fix some bugs related to parsing of band character names

 * added option --case

 * add line to postscript for explicit white background
   (current versions of ‘evince’ were drawing black bg)

 * add optimization options in Makefile (can be important even here!)


Version 0.1 (2013-05-31): initial release


## Local Variables:
## mode: text
## End:
##
## Time-stamp: <2015-09-14 17:56:14 assman@faepop23.tu-graz.ac.at>


prima.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

## prima.py -- generate band-character plot using Maurits Haverkort's
##             SpaghettiPrimavera
##
## Copyright 2013-2015 Elias Assmann <elias.assmann@gmail.com>

## prima.py version 0.3
##
## $Id: prima.py 618 2015-09-14 15:58:11Z assmann $

##
##	Vom Eise befreit sind Strom und Bäche
##	Durch des Frühlings holden, belebenden Blick,
##	Im Tale grünet Hoffnungsglück;
##	Der alte Winter, in seiner Schwäche,
##	Zog sich in rauhe Berge zurück.
##

from sppv import spaghettiprimavera, sppv_data as sppv

import os, sys, re, getopt, numpy, collections, webcolors, contextlib, \
  traceback, subprocess, warnings

def svn_rev():
    rev = "$Rev: 618 $"
    try:
        return '-r' + re.search('(\d+)', rev).group()
    except:
        return ''

__version__ = "0.3" + svn_rev()
sppv.version[:] = __version__

def carp(thing):
    if verbose and isinstance(thing, Exception):
        traceback.print_exc()
    else:
        if isinstance(thing, tuple):
            thing = ' '.join(str(x) for x in thing)
        print >>sys.stderr, "prima.py:", thing

def croak(thing, status=1):
    carp(thing)
    sys.exit(status)

@contextlib.contextmanager
def open_w_error(filename, mode="r"):
    try:
        f = open(filename, mode)
    except IOError, err:
        yield None, err
    else:
        try:
            yield f, None
        finally:
            f.close()

def find_reverse(lefi, string):
    """Find first line containing ‘string’ in ‘lefi’, searching backwards.
    ‘lefi’ must be seekable."""

    pos = lefi.tell()
    line  = ''
    while pos > 0:
        pos -= 1
        lefi.seek(pos)

        c = lefi.read(1)
        if c=='\n':
            if string in line:
                return line

            line = ''
        else:
            line = c + line

    if string in line: return line
    else:              return None

### The command line will be printed as a comment in the PS ###
sppv.cmdline[:] = ' '.join(sys.argv)

### Declare option related variables/functions ###
bandfile = None
def bandname(x):
    global bandfile
    print>>log, 'set klist file = ‘'       +x+'’'
    bandfile            = x
    sppv.bandname[:]    = x
def efermi(x):
    print>>log, 'set Fermi energy = ‘'     +x+'’ Ryd'
    sppv.efermi         = x
def emin(x):
    print>>log, 'set bottom energy = ‘'    +x+'’ eV'
    sppv.emin           = x
def emax(x):
    print>>log, 'set top energy = ‘'       +x+'’ eV'
    sppv.emax           = x
def majortics(x):
    print>>log, 'set major tics every = ‘' +x+'’ eV'
    sppv.majorticks     = x
def minortics(x):
    print>>log, 'set minor tics = ‘'       +x+'’ per major tic interval'
    sppv.minorticks     = x
def xsize(x):
    print>>log, 'set x-size = ‘'           +x+'’ pt'
    sppv.xsize          = x
def ysize(x):
    print>>log, 'set y-size = ‘'           +x+'’ pt'
    sppv.ysize          = x
def textsize(x):
    print>>log, 'set font size = ‘'        +x+'’ pt'
    sppv.textsize       = x
def fontname(x):
    print>>log, 'set font = ‘'             +x+'’'
    sppv.fontname[:]    = x
def legend(x):
    print>>log, 'set write legend'
    sppv.writelegend    = True
def nolegend(x):
    print>>log, 'unset legend'
    sppv.writelegend    = False
def noklines(x):
    print>>log, 'unset k-lines'
    sppv.writeklines    = False
def klines(x):
    print>>log, 'set draw k-lines'
    sppv.writeklines    = True
def noklabels(x):
    print>>log, 'unset k-labels'
    sppv.writeklabels   = False
def klabels(x):
    print>>log, 'set write k-labels'
    sppv.writeklabels   = True
def nofermi(x):
    print>>log, 'unset Fermi energy line'
    sppv.writefermiline = False
def fermi(x):
    print>>log, 'set draw Fermi energy line'
    sppv.writefermiline = True
def axes_thk(x):
    print>>log, 'set axis thickness = ‘'+x+'’ pt'
    sppv.axesthickness  = x
def base_thk(x):
    print>>log, 'set line base thickness = ‘'+x+'’ pt'
    sppv.basethickness  = x

qtlfile  = None
qtlfile1 = None
def qtlname(x):
    global qtlfile, qtlfile1
    qtlfile1        = qtlfile
    qtlfile         = x
    sppv.qtlname[:] = x
    print>>log, 'set qtl file = ‘'+x+'’'

def print_version(x):
    print 'prima.py version ' + __version__

    sys.exit()

def print_help(x):
    print "USAGE: prima.py [OPTIONS] [IATM:IORB:R,G,B:THCK:LEG ...]"
    print "       (generate plot)"
    print "   OR: prima.py [-q QTL_BAND] [-s STRUCT]"
    print "       (print available atom and orbital names)"
    print
    print "   IATM: atom number(s) or name(s) N[&M...]"
    print "   IORB: orbital character number(s) or name(s) N[&M...]"
    print "         default: tot"
    print "  R,G,B: color"
    print "         default: 0,0,0 (black)"
    print "   THCK: line thickness increment"
    print "         default: 0 (no “fat bands”)"
    print "    LEG: legend entry"
    print "         default: IATM-IORB (if -L option given)"
    print
    print "OPTIONS:"

    for o in [o for o in prima_options if o[3] is not None]:
        print "\t-" + o[0] + ", --" + o[1] + "\t" + o[3]
    for o in [o for o in sppv_options  if o[3] is not None]:
        print "\t-" + o[0] + ", --" + o[1] + "\t" + o[3]
    print
    print "Boolean options as listed negate the defaults; long options to reinstate"
    print "the defaults (e.g. ‘--no-legend’, ‘--k-labels’) are also valid."

    sys.exit()

spin=''
def do_up(x):
    global spin
    spin='up'
    print>>log, 'set spin = ‘'+spin+'’'
def do_dn(x):
    global spin
    spin='dn'
    print>>log, 'set spin = ‘'+spin+'’'
def do_updn(x):
    global spin
    spin='updn'
    print>>log, 'set spin = ‘'+spin+'’'
def no_spin(x):
    global spin
    spin=''

class nofile(object):
    def write(self, string):
        pass

verbose = False
log = nofile()
def do_verbose(x):
    global verbose, log
    verbose = True
    log = sys.stderr
    print>>log, 'set verbose'
def no_verbose(x):
    global verbose, log
    verbose = False
    print>>log, 'unset verbose'
    log = nofile()

outsuffix = None
def suffixname(x):
    global outsuffix
    outsuffix = x
    print>>log, 'set output suffix = ‘'+x+'’'
    print>>log

outfile = None
def outname(x):
    global outfile, outsuffix
    outfile = x
    outsuffix = None           # delete outsuffix for explicit outfile
    print>>log, 'set output file = ‘'+x+'’'

structfile = None
def structname(x):
    global structfile
    structfile = x
    print>>log, 'set struct = ‘'+x+'’'

scffile = None
def scfname(x):
    global scffile
    scffile = x
    print>>log, 'set scf file = ‘'+x+'’'

case = os.getcwd().split(os.sep)[-1]
def casename(x):
    global case
    case = x
    print>>log, 'set case = ‘'+x+'’'

def confname(x): # this is a no-op because config files are special
    # but it is a convenient place to verbosely echo config files
    print>>log, 'config file ‘'+x+'’read'
    

sppv_options = [
    ('f:', 'case=',           casename,     '\tdefault: $(basename $PWD)'),
    ('q:', 'qtl-file=',       qtlname,      'default: ‘case’.qtl[sp][_band]'),
    ('k:', 'klist-file=',     bandname,     'default: ‘case’.klist_band'),
    ('s:', 'struct-file=',    structname,   'default: ‘case’.struct\t\t[optional]'),
    ('c:', 'scf-file=',       scfname,      'default: ‘case’.scf\t\t[optional, for Fermi enerfy]'),
    ('F:', 'fermi-energy=',   efermi,       'Fermi energy in Rydberg; default: from scf'),
    ('e:', 'emin=',           emin,         '\tmin plotted energy in eV; default: -5'),
    ('E:', 'emax=',           emax,         '\tmax plotted energy in eV; default: +5'),
    ('t:', 'minor-tics=',     minortics,    '# minor tics per major interval; default: 4'),
    ('T:', 'major-tics=',     majortics,    'major tic distance in eV; default: 1'),
    ('X:', 'x-size=',         xsize,        '\tplot (not bbox) width  [pt]; default: 500'),
    ('Y:', 'y-size=',         ysize,        '\tplot (not bbox) height [pt]; default: 700'),
    ('S:', 'font-size=',      textsize,     'font size [pt?]; default: 12'),
    ('O:', 'font-name=',      fontname,     'font name; default: Times-Roman; Greek is in Symbol'),
    ('D:', 'base-thickness=', base_thk,     'base line thickness [pt?]; default: 0.1'),
    ('A:', 'axes-thickness=', axes_thk,     'line thickness for axes [pt?]; default: 0.5'),
    ('L',  'legend',          legend,       '\twrite legend'),
    (None, 'no-legend',       nolegend,     None),
    ('K',  'no-k-labels',     noklabels,    'omit labels for k-points'),
    (None, 'k-labels',        klabels,      None),
    ('G',  'no-k-lines',      noklines,     'omit vertical lines for labeled k-points'),
    (None, 'k-lines',         klines,       None),
    ('0',  'no-fermi-line',   nofermi,      'omit horizontal line for Fermi energy'),
    (None, 'fermi-line',      fermi,        None),
]

prima_options = [
    ('u',  'up',              do_up,         '\tsp: up'),
    ('d',  'dn',              do_dn,         '\tsp: down'),
    ('x',  'updn',            do_updn,       '\tmix spins'),
    ('N', 'no-spin',          no_spin,       '\tnon-spin-polarized'),
    ('o:', 'out-file=',       outname,       'send output here instead of STDOUT'),
    ('a:', 'out-suffix=',     suffixname,    'send output to $case$suffix instead of STDOUT'),
    ('C:', 'config-file=',    confname,      'read options from here'),
    ('h',  'help',            print_help,    '\tthis message'),
    ('v',  'verbose',         do_verbose,    '\tprint additional info to STDERR'),
    (None, 'no-verbose',      do_verbose,    None),
    ('V',  'version',         print_version, '\tversion info'),
]

### Parse options ###
opt_dict = {}
shopts = ""
longopts = []

for o in sppv_options + prima_options:
    l = o[1]
    if l[-1] == '=': l = l[0:-1]

    opt_dict[l] = o
    longopts += [o[1]]

    if o[0] is not None:
        opt_dict[o[0][0]] = o
        shopts   +=  o[0]

print>>log, 'Parsing option arguments …'

try:
    (opts, args) = getopt.gnu_getopt(sys.argv[1:], shopts, longopts)
except getopt.GetoptError as e:
    croak(e)

## First, look for and read config files
xopts = []
skipre = re.compile('\s*(#.*)?$')
for (o, a) in opts:
    if o not in ['-C', '--config-file']: continue

    try:
        c = open(a, 'U')

        for line in c:
            if skipre.match(line): continue

            s = line.rstrip('\n').split(None, 1)

            lo = s[0]

            if len(s) > 1:
                la = s[1]
            else:
                la = ''

            xopts += [(lo, la)]

    except IOError as e:
        croak(('error reading config file:', e))

opts = xopts + opts
    
## Now for the other options
stripdash = re.compile('^--?')

for (o, a) in opts:
    o = stripdash.sub('', o)
    opt_dict[o][2](a)

### Set defaults ###

## This is done here because it relies on ‘case’ and some other
## options being set

if args:
    ## set efermi from scf unless it has been set
    if not [x for x in opts if x[0] in ['-F', '--fermi-energy']]:
        try:
            if scffile is None: scffile = case + '.scf'

            print>>log, 'no -F option, will try to find Fermi energy in ‘'\
                         +scffile+'’'

            with open(scffile, 'U') as scf:
                scf.seek(0, os.SEEK_END)

                line = find_reverse(scf, ':FER')
                if spin=='up':
                    line = find_reverse(scf, ':FER')

                if line is None:
                    raise Exception('Fermi energy not found in ‘' +
                                    scffile + '’')

                sppv.efermi = float(line.split()[-1])

        except Exception as e:
            croak(('error trying to fetch Fermi energy:', e))

    print>>log, 'Fermi energy is EF =', sppv.efermi, 'Ryd'

## set defaults for qtlfile, bandname, struct unless set
print>>log
print>>log, 'Files used:'
if qtlfile    is None:
    s = spin
    t = 'up'
    if spin == 'updn': s = 'dn'

    qtlfile      = case + '.qtl' + s + '_band'
    qtlfile1     = case + '.qtl' + t + '_band'
    if not os.path.exists(qtlfile):
        qtlfile  = case + '.qtl' + s
        qtlfile1 = case + '.qtl' + t

    sppv.qtlname[:] = qtlfile

    if spin == 'updn':
        print>>log, '   qtl1= ‘'+qtlfile1+'’'
    print>>log,     '   qtl = ‘'+qtlfile +'’'
elif spin=='updn' and  qtlfile1 is None:
    croak('need 0 or 2 qtl files in --updn mode, not 1')

if bandfile   is None:
    bandfile         = case + '.klist_band'
    sppv.bandname[:] = bandfile
    print>>log, '   klist_band = ‘'+bandfile+'’'
if structfile is None:
    structfile = case + '.struct'
    print>>log, '   struct = ‘'+structfile+'’'

## set outfile from outsuffix if set
if outsuffix is not None:
    outfile = case + outsuffix

if outfile is not None:
    print>>log, '   output to ‘'+outfile+'’'

### Read QTL file for orbital character info ###
print>>log
print>>log, 'Reading qtl file …'
try:
    qtl     = open(qtlfile, 'U')
except IOError as e:
    croak(('error opening qtl file:', e))

jatomre = re.compile(' *JATOM *.* (tot.*)')
endre   = re.compile(' *BAND')

have_orbs = []
for line in qtl:
    m = jatomre.match(line)
    if m:
        print>>log
        print>>log, 'Next atom'
        o = {}

        for i,name in enumerate(m.group(1).split(',')):
            name = name.strip()

            if name == '': break

            name = {
                '0': 'S',
                '1': 'P',
                '2': 'D',
                '3': 'F'
                }.get(name, name).lower()
            o[name] = i
            print>>log, '   found orbital', name, '=>', i

        have_orbs.append(o)

    if endre.match(line): break

qtl.close()

num_atoms = len(have_orbs)

have_orbs.append({'tot' : 0})   # interstitial


### Read struct file for atom info ###
# For each key (atom name, symbol, atomic number), have_atoms will
# contain a set of atom indices that correspond to that key
have_atoms = { 'all': set() }

try:
    print>>log
    print>>log, 'Reading struct file …'
    struct = open(structfile, 'U')

    atomre = re.compile('(.*?) *NPT=')
    symre  = re.compile('[a-zA-Z]+')
    zre    = re.compile('Z: *([\d.]+)')

    iat = 1
    for line in struct:
        m = atomre.match(line)
        if m:
            # populate have_atoms: atom name
            aname = m.group(1)
            try:             have_atoms[aname.lower()].add(iat)
            except KeyError: have_atoms[aname.lower()] =  {iat}

            # atom symbol
            m = symre.match(aname)
            if m is None: continue

            asym = m.group()
            try:             have_atoms[asym.lower()].add(iat)
            except KeyError: have_atoms[asym.lower()] =  {iat}

            # atomic number
            m = zre.search(line)
            if m is None: continue

            az = "z%g" % float(m.group(1))
            try:             have_atoms[az].add(iat)
            except KeyError: have_atoms[az] =  {iat}

            # 'all'
            have_atoms['all'].add(iat)

            print>>log, '   found atom', iat, '<=', aname, asym, az

            iat += 1

    have_atoms['istl'] = {iat}

    if iat-1 != num_atoms:
        croak('struct (%i atoms) and qtl (%i atoms) files inconsistent'
              %(iat-1, num_atoms))

    struct.close()
except IOError: pass                # FIXME: this could be more robust

### Parse orbital character specs ###
atms = []
orbs = []
clrs = []
thks = []

legnames  = []
legcolors = []

combinator = re.compile('\+\+|&')
multiplier = re.compile('x|×')

def parsecolor(cc):
    if multiplier.search(cc):
        (fact, cc) = multiplier.split(cc, 1)
        fact = float(fact)
    else:
        fact = 1.

    cc = cc.strip()

    try:
        c = numpy.array(map(float, cc.split(',')))

        if c.size == 1: c = c.repeat(3)
        if c.size != 3: raise StandardError()

        return c * fact
    except ValueError: pass

    try:
        c = webcolors.hex_to_rgb(cc)
        return numpy.array(c)/255. * fact
    except ValueError: pass

    c = webcolors.name_to_rgb(cc)
    return numpy.array(c)/255. * fact

print>>log
print>>log, 'Parsing non-option arguments …'

for a in args:
    tok = a.split(':', 4)

    # legend entry provided, activate legend
    if len(tok) > 4: legend(True)

    tok += [''] * (5 - len(tok))

    print>>log, '   found tokens', tok,

    (aa, oo, cc, t, l) = tok[:5]

    # defaults
    if oo == '': oo = 'tot'
    if cc == '': cc = '0,0,0'
    if t  == '': t  = '0'
    if l  == '': l  = tok[0]+'-'+oo
    
    aa = aa.lower()
    oo = oo.lower()

    # transform color arg to array
    try:
        c = parsecolor(cc)
    except StandardError:
        croak(('invalid color spec:', cc))

    # store legend entry -- there should be one entry for each command
    # line arg, not for each expanded atom/orbital
    if l:
        legnames .append(l)
        legcolors.append(c)

    print>>log, '=> [', aa, oo, c, t, l, ']'

    # now expand atom / orbital args
    aa = collections.deque(combinator.split(aa))

    upsym = ['-up', '-↑', '↑']
    dnsym = ['-dn', '-↓', '↓']
    udsym = ['-ud', '-↓↑', '↑↓']
    spinre = re.compile('(.*)' + '(' + '|'.join(upsym+dnsym+udsym) + ')')
    mult = 2 if spin=='updn' else 1
    while aa:
        a = aa.popleft()

        add = 0
        sp  = ''
        if spin=='updn' and isinstance(a, str):
            m = spinre.match(a)
            if m:
                (a, sp) = m.groups()
            else:                        # empty spin part means up+dn
                sp = udsym[0]

        if sp in dnsym: add = 1

        if sp in udsym:
            aa += [a+upsym[0], a+dnsym[0]]
            continue

        try:
            iat = int(a)-1
        except ValueError:
            a = a.strip()
            try:
                xa = [str(i)+sp for i in have_atoms[a]]
                aa.extendleft(xa)
                continue
            except KeyError:
                croak('atom ‘' + a + '’ not found')

        for o in combinator.split(oo):
            o = o.strip()

            try:
                iorb = int(o)-1

                if iorb < 0 or iorb >= len(have_orbs[iat]):
                    raise KeyError(iorb)
            except ValueError:
                try:
                    iorb = have_orbs[iat][o]
                except KeyError:
                    croak('orbital ‘' + o + '’ not found')
            except KeyError:
                croak('orbital ‘' + o + '’ not found')

            atms.append(1 + mult*iat + add)      # +1 for Fortran indices
            orbs.append(1 + iorb)
            clrs.append(c)
            thks.append(float(t))

            print>>log, \
                '      adding atom %i, orb %i, color %s, thickness %g' \
                %(1+2*iat+add, 1 + iorb, c, float(t))

print>>log
print>>log, "Atoms   :", atms
print>>log, "Orbitals:", orbs
print>>log, "Colors  :", clrs
print>>log, "Thickn's:", thks

sppv.orbcolors = numpy.array(clrs).T
sppv.atomsorbs = numpy.array((atms, orbs))
sppv.orbthicks = thks

if sppv.writelegend:
    # propagate legend info to Fortran
    sppv.legcolors = numpy.array(legcolors).T

    # FIXME: the ‘+ ' '’ seems to prevent a segfault if a single
    # legend entry of one character is given
    sppv.legend = ''.join(legnames) + ' '

    sppv.legentries = numpy.cumsum(map(len, legnames))

## no arguments: print available characters
if not atms:
    print>>log
    print>>log, 'No non-option arguments: print available characters'

    print 'prima.py', __version__, 'using', qtlfile, 'and', structfile
    print
    for a,o in enumerate(have_orbs, start=1):
        print a, ':',
        for io in sorted(o.iteritems(), cmp=lambda a,b: a[1]-b[1]):
            print repr(io[0]) + ':' + repr(io[1]+1) + ',',
        else: print
    print
    for atn,iat in sorted(have_atoms.iteritems()):
        atn = atn[0].upper() + atn[1:]
        print atn, '=>', ','.join(map(repr, iat))

    print>>log
    print>>log, 'All done.'
    sys.exit()

### Real Work ###

## In ‘updn’ mode, open a pipe to ‘paste’ the two qtl files
if spin=='updn':
    print>>log
    print>>log, 'Opening pipe to paste together ‘up’ and ‘dn’ qtl files …'

    warnings.simplefilter('ignore')
    qtlpipe = os.tempnam(None, 'udqtl')
    warnings.resetwarnings()
    try:
        os.mkfifo(qtlpipe)
        sppv.qtlname[:] = qtlpipe

        if os.fork()==0: # child
            # open() will block until fifo opened for reading
            with open(qtlpipe, 'w') as qtl, \
                 open(qtlfile1) as q1, open(qtlfile) as q2:
                # header
                natre = re.compile('NAT\s*=\s*(\d+)')
                l1=''
                while 'BAND' not in l1:
                    print>>qtl, l1,

                    if 'JATOM' in l1:
                        print>>qtl, l2,

                    m = natre.match(l1)
                    if m:
                        if int(m.group(1)) != num_atoms:
                            croak(('inconsistent number of atoms in ‘%s’ ' +
                                   '(%i atoms instead of %i)'
                                   %(qtlfile1, int(m.group(1)), num_atoms)))

                        m = natre.match(l2)
                        if int(m.group(1)) != num_atoms:
                            croak(('inconsistent number of atoms in ‘%s’ ' +
                                   '(%i atoms instead of %i)'
                                   %(qtlfile2, int(m.group(1)), num_atoms)))
                    # get new lines
                    l1 = q1.readline()
                    l2 = q2.readline()

                # header done, l1 and l2 contain “BAND 1”; print extra
                # line for interstitial
                print>>qtl, 'JATOM X interstitial'
                print>>qtl, l1,

                for l1 in q1:
                    l2 = q2.readline()
                    print>>qtl, l1,

                    if 'BAND' not in l1:
                        print>>qtl, l2,

            sys.exit()
    except EnvironmentError as e:
        croak(('error opening pipe for --updn qtl:', e))
    finally:
        os.unlink(qtlpipe)

    print>>log, 'Using named pipe ‘'+qtlpipe+'’as qtl file'

## Redirect STDOUT if necessary -- need to use low-level I/O to
## propagate change to Fortran
if outfile not in [None, '-']:
    print>>log
    print>>log, 'Redirecting STDOUT for Fortran'
    os.close(1)
    os.umask(~0644)
    # should open on 1 (fingers crossed)
    fid = os.open(outfile, os.O_WRONLY|os.O_CREAT|os.O_TRUNC)
    if fid != 1:
        croak(('Redirecting STDOUT did not work -- opened on', fid,
               'instead of 1'))

print>>log
print>>log, 'Handing over to Fortran …'

spaghettiprimavera()

print>>log
print>>log, 'All done.'


## Time-stamp: <2015-09-22 16:46:36 assman@faepop36.tu-graz.ac.at>
